pkg/server/server_engine_health.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/storage/engine"
+	"github.com/cockroachdb/cockroach/pkg/util/envutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+var engineHealthCheckInterval = envutil.EnvOrDefaultDuration("COCKROACH_ENGINE_HEALTH_CHECK_INTERVAL", 5*time.Second)
+
+func startAssertEngineHealth(ctx context.Context, stopper *stop.Stopper, engines []engine.Engine) {
+	stopper.RunWorker(ctx, func(ctx context.Context) {
+		t := timeutil.NewTimer()
+		t.Reset(0)
+
+		for {
+			select {
+			case <-t.C:
+				t.Read = true
+				t.Reset(engineHealthCheckInterval)
+				assertLoggingHealth(ctx, engineHealthCheckInterval)
+				assertEngineHealth(ctx, engines, engineHealthCheckInterval)
+			case <-stopper.ShouldQuiesce():
+				return
+			}
+		}
+	})
+}
+
+// Don't change this string without also changing the disk-stall-detection
+// roachtest which greps for it.
+const diskStallDetected = "disk stall detected"
+
+func assertEngineHealth(ctx context.Context, engines []engine.Engine, maxDuration time.Duration) {
+	for _, eng := range engines {
+		func() {
+			t := time.AfterFunc(maxDuration, func() {
+				log.Shout(ctx, log.Severity_FATAL, fmt.Sprintf(
```

Comment:
here and below: this is likely to not-work when crash reporting is enabled. In that case, a FATAL severity will cause things to happen (including disk access) before the call to os.Exit.

So, either:

- ensure the tests for this occur with crash reporting is enabled, or
- modify the code further so that there is a guaranteed call to os.Exit even if the log.Shout never terminates. For example

```go
time.AfterFunc(maxDuration, func ...)
time.Afterfunc(maxDuration + 1 * time.Second, func() { os.Exit(...) })
```

Reply:
Ah, I had reviewed `log.Fatal` and saw that it had a case for this and assumed things were handled already. I'll make sure to adopt your suggestion.